src/subshell/common.c

```diff
@@ -679,37 +643,55 @@ read_command_line_buffer (gboolean test_mode)
          *
          * See #4625, in particular #issuecomment-3425779646. */
         if (FD_ISSET (mc_global.tty.subshell_pty, &read_set))
-            flush_subshell (0, test_mode ? QUIETLY : VISIBLY);
+            flush_subshell (0, how);
 
         if (FD_ISSET (command_buffer_pipe[READ], &read_set))
             break;
     }
 
-    bytes =
-        read (command_buffer_pipe[READ], subshell_command_buffer, sizeof (subshell_command_buffer));
+    *bytes_read = read (command_buffer_pipe[READ], buffer, buffer_size);
+
     // FIXME: what about bytes < 0?
-    if (bytes == 0 || (size_t) bytes == sizeof (subshell_command_buffer))
+    if (*bytes_read == 0 || (size_t) *bytes_read == buffer_size)
         return FALSE;
 
-    command_buffer_char_length = (size_t) bytes - 1;
-    subshell_command_buffer[command_buffer_char_length] = '\0';
-    command_buffer_length = str_length (subshell_command_buffer);
+    return TRUE;
+}
 
-    // get cursor position
-    write_all (mc_global.tty.subshell_pty, ESC_STR SHELL_CURSOR_KEYBINDING,
-               sizeof (ESC_STR SHELL_CURSOR_KEYBINDING) - 1);
+/* --------------------------------------------------------------------------------------------- */
+/* Get the contents of the current subshell command line buffer, and */
+/* transfer the contents to the panel command prompt. */
 
-    subshell_prompt_timer.tv_sec = 1;
-    subshell_prompt_timer.tv_usec = 0;
+static gboolean
+read_command_line_buffer (gboolean test_mode)
+{
+    char subshell_command_buffer[BUF_LARGE];
+    char subshell_cursor_buffer[BUF_SMALL];
 
-    while (TRUE)
-    {
-        FD_ZERO (&read_set);
-        FD_SET (command_buffer_pipe[READ], &read_set);
-        FD_SET (mc_global.tty.subshell_pty, &read_set);
+    fd_set read_set;
+    ssize_t bytes;
+    struct timeval subshell_prompt_timer = {
+        .tv_sec = 0,
+        .tv_usec = 0,
+    };
+    int bash_version;
+    int cursor_position;
+    int rc;
 
-        rc = select (maxfdp + 1, &read_set, NULL, NULL, &subshell_prompt_timer);
+    if (!use_persistent_buffer)
+        return TRUE;
 
+    FD_ZERO (&read_set);
+    FD_SET (command_buffer_pipe[READ], &read_set);
+
+    const int nfds = MAX (command_buffer_pipe[READ], mc_global.tty.subshell_pty) + 1;
+
+    /* First, flush the command buffer pipe. This pipe shouldn't be written
```

Comment:
this could be also de-duplicated, at the cost of adding a flag to the reader function.
not sure it's worth it at this point.